### DISCUSSION TITLE
Getting a unique result from Select API ORM query

### DISCUSSION THREAD

**Discussion by willnewton** [sqlalchemy/sqlalchemy#11672] [Usage Questions] [answered]
Hi,

I have an existing codebase that uses the Query API to return a set of objects. I'm trying to convert it to use the Select API but running into an issue.

One of the objects involved is quite complex and has a field which is an array of enum and also has a collection of objects attached.

If I try to naively return the result set:

```
session.execute(query).unique().all()
```

I get the error:

```
The unique() method must be invoked on this Result, as it contains results that include joined eager loads against collections
```

However if I try to add unique() I get:

```
Can't apply uniqueness to row tuple containing value of type ARRAY(ENUM(...)); this datatype produces non-hashable values
```

What's the best way to resolve this so the behaviour is the same as with the Query API?

**Comment by zzzeek** [comment, marked as answer]
legacy Query is using unique() here as well, however its unique filter is modified to use the id() function on the object, rather than raising the "cant apply uniqueness" error.    If the returned result is an ARRAY[ENUM], then id() is going to return a different value each time and unique() is not getting you anything at all.

So I would normally seek to not use the unique() function here, it would only be required if you are using collection related joined eager loading, which similarly would not be reliable here since these rows arent actually being uniqued.

There's other options here:

* Use the [as_tuple](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.ARRAY.params.as_tuple) parameter on your ARRAY which should make the values hashable
* Use a [TypeDecorator](https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator) around your ARRAY type to return some object that's hashable (if the as_tuple is not enough)
* Supply a filtering function to the [result.unique()](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.unique) method.   this receives the row and should return any kind of hashable object that represents the unique value of the row

**Reply by willnewton** [reply]
Thanks, that's super helpful!
